### ISSUE TITLE
Run `refPath` and `ref` functions on subdocument being populated rather than top level doc

### ISSUE THREAD

**Issue by vkarpov15** [Automattic/mongoose#8469]
Re: #8452. This makes it exceedingly difficult to use `ref` functions as a replacement for `refPath`, and makes it hard to populate conditionally based on the contents of the subdocument.

```javascript
    const ImageSchema = Schema({ imageName: String });
    const Image = db.model('gh8452_Image', ImageSchema);

    const TextSchema = Schema({ textName: String });
    const Text = db.model('gh8452_Text', TextSchema);

    const opts = { _id: false };
    const ItemSchema = Schema({ objectType: String }, opts);
    const ItemSchemaA = Schema({
      data: { 
        type: ObjectId,
        refPath: function() { return 'list.objectType'; } // `this` will be the top-level `Example` doc
      },
      objectType: String,
    }, opts);
    const ItemSchemaB = Schema({
      data: { sourceId: Number },
      objectType: String,
    }, opts);

    const ExampleSchema = Schema({ test: String, list: [ItemSchema] });
    ExampleSchema.path('list').discriminator('gh8452_ExtendA', ItemSchemaA);
    ExampleSchema.path('list').discriminator('gh8452_ExtendB', ItemSchemaB);
    const Example = db.model('gh8452_Example', ExampleSchema);
```

**Comment by dennisleung** [comment]
is there any solution for this? ...

**Comment by vkarpov15** [comment]
@dennisleung not currently. We're working on a fix for v6.0.0.

**Comment by vkarpov15** [comment]
`ref` is done in e141409, did some work for `refPath` in https://github.com/Automattic/mongoose/commit/6179267b61bb34d0b0b95ba36e35103db5ae2b7f.

Worth noting: I think we should treat `refPath` as being relative to the top-level document rather than the subdocument. There is a benefit to calling `refPath` with the subdocument as context (say, switching `refPath` based on the type of `this`), but Mongoose should treat the returned path as relative to the top level document, so users can reference top-level properties.
